File: 078_subsets/Solution1/Solution.go
package main

import "fmt"

func subsets(nums []int) [][]int {
    arrLen := len(nums)
    var results [][]int

    for i := 0; i <= arrLen; i++ {
        var path []int
        dfs(nums, 0, arrLen, i, &results, path)
    }

    return results
}

func dfs(nums []int, start, arrLen, steps int, results *[][]int, path []int) {
    if steps == 0 {
        *results = append(*results, append([]int{}, path...))
        return
    }

    for i := start; i < arrLen; i++ {
        path = append(path, nums[i])
        dfs(nums, i + 1, arrLen, steps - 1, results, path)
        path = path[:len(path) - 1]
    }
}

func main() {
    nums := []int{1, 2, 3}
    results := subsets(nums)
    for _, result := range results {
        fmt.Println(result)
    }
}

File: 078_subsets/Solution1/Solution.py
class Solution:
    def subsets(self, nums: list) -> list:
        results = []
        arr_len = len(nums)

        def dfs(start: int, steps: int, path: list):
            if steps == 0:
                results.append(path[:])
                return

            for i in range(start, arr_len):
                path.append(nums[i])
                dfs(i + 1, steps - 1, path)
                path.pop()

        for i in range(arr_len + 1):
            dfs(0, i, [])

        return results


    def subsets2(self, nums: list) -> list:
        results = []
        arr_len = len(nums)

        def dfs(idx: int, tmp_list: list):
            results.append(tmp_list)
            for i in range(idx, arr_len):
                dfs(i + 1, tmp_list + [nums[i]])

        dfs(0, [])
        return results


if __name__ == '__main__':
    nums = [1, 2, 3]
    print(Solution().subsets(nums))

